Exception thrown for Index shard.

[IndexShard.txt](https://github.com/gchq/stroom/files/8685921/IndexShard.txt)

admin```
this writer hit an unrecoverable error; cannot complete commit 
java.lang.IllegalStateException: this writer hit an unrecoverable error; cannot complete commit
	at org.apache.lucene.index.IndexWriter.finishCommit(IndexWriter.java:2985)
	at org.apache.lucene.index.IndexWriter.commitInternal(IndexWriter.java:2970)
	at org.apache.lucene.index.IndexWriter.commit(IndexWriter.java:2930)
	at stroom.index.impl.IndexShardWriterImpl.flush(IndexShardWriterImpl.java:284)
	at stroom.index.impl.IndexShardWriterCacheImpl.lambda$flush$14(IndexShardWriterCacheImpl.java:396)
	at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$2(TaskContextFactoryImpl.java:255)
	at stroom.util.logging.LocationAwareLambdaLogger.logDurationIfDebugEnabled(LocationAwareLambdaLogger.java:156)
	at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$4(TaskContextFactoryImpl.java:255)
	at stroom.util.pipeline.scope.PipelineScopeRunnable.scopeResult(PipelineScopeRunnable.java:38)
	at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$5(TaskContextFactoryImpl.java:252)
	at stroom.security.impl.SecurityContextImpl.asUserResult(SecurityContextImpl.java:232)
	at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$9(TaskContextFactoryImpl.java:252)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1764)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: java.io.IOException: null: NIOFSIndexInput(path=&#34;/stroom/volumes/xxxxxxx_index/index/3e72379d-403f-4757-88d5-bfae4d724b55/2022-05/451/_47.fdt&#34;)
	at org.apache.lucene.store.NIOFSDirectory$NIOFSIndexInput.readInternal(NIOFSDirectory.java:189)
	at org.apache.lucene.store.BufferedIndexInput.readBytes(BufferedIndexInput.java:160)
	at org.apache.lucene.store.BufferedIndexInput.readBytes(BufferedIndexInput.java:116)
	at org.apache.lucene.store.BufferedChecksumIndexInput.readBytes(BufferedChecksumIndexInput.java:49)
	at org.apache.lucene.store.DataInput.readBytes(DataInput.java:87)
	at org.apache.lucene.store.DataInput.skipBytes(DataInput.java:350)
	at org.apache.lucene.store.ChecksumIndexInput.seek(ChecksumIndexInput.java:52)
	at org.apache.lucene.codecs.CodecUtil.checksumEntireFile(CodecUtil.java:450)
	at org.apache.lucene.codecs.compressing.CompressingStoredFieldsReader.checkIntegrity(CompressingStoredFieldsReader.java:669)
	at org.apache.lucene.codecs.compressing.CompressingStoredFieldsWriter.merge(CompressingStoredFieldsWriter.java:534)
	at org.apache.lucene.index.SegmentMerger.mergeFields(SegmentMerger.java:177)
	at org.apache.lucene.index.SegmentMerger.merge(SegmentMerger.java:83)
	at org.apache.lucene.index.IndexWriter.mergeMiddle(IndexWriter.java:4075)
	at org.apache.lucene.index.IndexWriter.merge(IndexWriter.java:3655)
	at org.apache.lucene.index.ConcurrentMergeScheduler.doMerge(ConcurrentMergeScheduler.java:588)
	at org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:626)
Caused by: java.nio.channels.ClosedChannelException: null
	at java.base/sun.nio.ch.FileChannelImpl.ensureOpen(FileChannelImpl.java:156)
	at java.base/sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:790)
	at org.apache.lucene.store.NIOFSDirectory$NIOFSIndexInput.readInternal(NIOFSDirectory.java:179)
	... 15 common frames omitted
```

--- a/stroom-index/stroom-index-impl/src/main/java/stroom/index/impl/IndexShardWriterImpl.java
+++ b/stroom-index/stroom-index-impl/src/main/java/stroom/index/impl/IndexShardWriterImpl.java
@@ -26,6 +26,7 @@ import stroom.util.NullSafe;
 import stroom.util.io.FileUtil;
 import stroom.util.logging.LambdaLogger;
 import stroom.util.logging.LambdaLoggerFactory;
+import stroom.util.logging.LogUtil;
 import stroom.util.logging.LoggerPrintStream;
 import stroom.util.shared.ModelStringUtil;
 import stroom.util.time.StroomDuration;
@@ -138,6 +139,7 @@ public class IndexShardWriterImpl implements IndexShardWriter {
 
         // Find the index shard dir.
         dir = IndexShardUtil.getIndexPath(indexShard);
+        LAMBDA_LOGGER.debug(() -> LogUtil.message("Creating index shard writer for dir {} {}", dir, this));
 
         // Make sure the index writer is primed with the necessary analysers.
         LAMBDA_LOGGER.debug(() -> "Updating field analysers");
@@ -152,12 +154,15 @@ public class IndexShardWriterImpl implements IndexShardWriter {
         final String path = FileUtil.getCanonicalPath(dir);
         if (Files.isDirectory(dir)) {
             final long count = FileUtil.count(dir);
+            LAMBDA_LOGGER.debug(() -> dir + " exists, file count: " + count + " " + this);
             if (count == 0) {
                 throw new IndexException("Unable to find any index shard data in directory: " + path);
             }
 
         } else {
             // Make sure the index hasn't been deleted.
+            LAMBDA_LOGGER.debug(() ->
+                    dir + " doesn't exist, doc count: " + indexShard.getDocumentCount() + " " + this);
             if (indexShard.getDocumentCount() > 0) {
                 throw new IndexException("Unable to find any index shard data in directory: " + path);
             }
@@ -166,7 +171,9 @@ public class IndexShardWriterImpl implements IndexShardWriter {
             try {
                 Files.createDirectories(dir);
             } catch (final IOException e) {
-                LAMBDA_LOGGER.error(e::getMessage, e);
+                LAMBDA_LOGGER.error(() ->
+                                buildErrorMessage("Error creating path " + dir, e),
+                        e);
                 throw new IndexException("Unable to create directories for new index in \"" + path + "\"");
             }
         }
@@ -193,6 +200,7 @@ public class IndexShardWriterImpl implements IndexShardWriter {
 
         // IndexWriter to use for adding data to the index.
         final IndexWriter indexWriter = new IndexWriter(directory, luceneIndexWriterConfig);
+        LOGGER.debug("Marking shard writer as open. {}", this);
         open.set(true);
 
         final LiveIndexWriterConfig liveIndexWriterConfig = indexWriter.getConfig();
@@ -208,7 +216,12 @@ public class IndexShardWriterImpl implements IndexShardWriter {
 
         if (indexShard.getDocumentCount() != numDocs) {
             LAMBDA_LOGGER.error(() ->
-                    "Mismatch document count. Index says " + numDocs + " DB says " + indexShard.getDocumentCount());
+                    "Mismatch document count. Index says "
+                            + numDocs
+                            + " DB says "
+                            + indexShard.getDocumentCount()
+                            + " "
+                            + this);
         }
 
         this.directory = directory;
@@ -239,7 +252,7 @@ public class IndexShardWriterImpl implements IndexShardWriter {
                     if (duration.compareTo(slowIndexWriteWarningThreshold.getDuration()) > 0) {
                         LAMBDA_LOGGER.warn(() ->
                                 "addDocument() - "
-                                        + this.toString()
+                                        + this
                                         + " took "
                                         + duration
                                         + " (Warning threshold: "
@@ -266,7 +279,8 @@ public class IndexShardWriterImpl implements IndexShardWriter {
                 // Add the field analyser.
                 final Analyzer analyzer = AnalyzerFactory.create(indexField.getAnalyzerType(),
                         indexField.isCaseSensitive());
-                LAMBDA_LOGGER.debug(() -> "Adding field analyser for: " + indexField.getFieldName());
+                LAMBDA_LOGGER.debug(() ->
+                        "Adding field analyser for: " + indexField.getFieldName() + " " + toString());
                 fieldAnalyzers.put(indexField.getFieldName(), analyzer);
             }
         }
@@ -284,7 +298,7 @@ public class IndexShardWriterImpl implements IndexShardWriter {
                 indexWriter.commit();
 
             } catch (final IOException | RuntimeException e) {
-                LAMBDA_LOGGER.error(e::getMessage, e);
+                LAMBDA_LOGGER.error(buildErrorMessage("Error while committing writer.", e), e);
 
             } finally {
                 // Update the shard info
@@ -315,6 +329,7 @@ public class IndexShardWriterImpl implements IndexShardWriter {
                         Thread.sleep(1000);
                     }
                 } catch (final InterruptedException e) {
+                    LAMBDA_LOGGER.debug(() -> "Interrupted waiting for docs to be added. " + toString());
                     // Interrupt this thread again.
                     Thread.currentThread().interrupt();
                 }
@@ -329,15 +344,14 @@ public class IndexShardWriterImpl implements IndexShardWriter {
                             directory.close();
                         }
                     } catch (final IOException | RuntimeException e) {
-                        LAMBDA_LOGGER.error(e::getMessage, e);
+                        LOGGER.error(buildErrorMessage("Error closing directory.", e), e);
                     }
 
                     open.set(false);
                 }
 
             } catch (final RuntimeException e) {
-                LAMBDA_LOGGER.error(e::getMessage, e);
-
+                LOGGER.error(buildErrorMessage("Error closing shard writer.", e), e);
             } finally {
                 // Update the shard info
                 updateShardInfo(startTime);
@@ -366,7 +380,7 @@ public class IndexShardWriterImpl implements IndexShardWriter {
 
             update(indexShardId, lastDocumentCount, lastCommitDurationMs, lastCommitMs, fileSize);
         } catch (final RuntimeException e) {
-            LAMBDA_LOGGER.error(e::getMessage, e);
+            LOGGER.error(buildErrorMessage("Error updating shard info.", e), e);
         }
     }
 
@@ -380,16 +394,16 @@ public class IndexShardWriterImpl implements IndexShardWriter {
                         try {
                             totalSize.getAndAdd(Files.size(file));
                         } catch (final IOException e) {
-                            LOGGER.trace(e.getMessage(), e);
+                            LOGGER.trace(buildErrorMessage("IO error getting file size.", e), e);
                         }
                     });
                 } catch (final IOException e) {
-                    LOGGER.trace(e.getMessage(), e);
+                    LOGGER.trace(buildErrorMessage("IO error getting file sizes.", e), e);
                 }
                 fileSize = totalSize.get();
             }
         } catch (final RuntimeException e) {
-            LAMBDA_LOGGER.debug(e::getMessage, e);
+            LOGGER.debug(buildErrorMessage("Error calculating file sizes.", e), e);
         }
         return fileSize;
     }
@@ -436,5 +450,14 @@ public class IndexShardWriterImpl implements IndexShardWriter {
     @Override
     public String toString() {
         return "(id=" + indexShardId + ")";
+    }
+
+    private String buildErrorMessage(final String message,
+                                     final Throwable throwable) {
+        return message
+                + " "
+                + NullSafe.getOrElse(throwable, Throwable::getMessage, "null")
+                + " "
+                + this;
     }
 }

--- a/stroom-index/stroom-index-impl/src/main/java/stroom/index/impl/IndexShardWriterImpl.java
+++ b/stroom-index/stroom-index-impl/src/main/java/stroom/index/impl/IndexShardWriterImpl.java
@@ -184,10 +184,10 @@ public class IndexShardWriterImpl implements IndexShardWriter {
         final PerFieldAnalyzerWrapper analyzerWrapper = new PerFieldAnalyzerWrapper(defaultAnalyzer, fieldAnalyzers);
         final IndexWriterConfig luceneIndexWriterConfig = new IndexWriterConfig(analyzerWrapper);
 
-        // In debug mode we do extra trace in LUCENE and we also count
+        // In trace mode we do extra trace in LUCENE and we also count
         // certain logging info like merge and flush
         // counts, so you can get this later using the trace method.
-        if (LOGGER.isDebugEnabled()) {
+        if (LOGGER.isTraceEnabled()) {
             final LoggerPrintStream loggerPrintStream = new LoggerPrintStream(LOGGER);
             for (final String term : LOG_WATCH_TERMS.values()) {
                 loggerPrintStream.addWatchTerm(term);

--- a/stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/ExtractionDecorator.java
+++ b/stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/ExtractionDecorator.java
@@ -38,7 +38,7 @@ import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
 import java.util.function.Supplier;
 import javax.inject.Provider;
 
@@ -232,7 +232,7 @@ public class ExtractionDecorator {
 
     @SuppressWarnings("unchecked")
     public CompletableFuture<Void> startExtraction(final TaskContext parentContext,
-                                                   final AtomicLong extractionCount,
+                                                   final LongAdder extractionCount,
                                                    final ErrorConsumer errorConsumer) {
         // Delay extraction if we are going to use one or more extraction pipelines.
         receivers.keySet().stream().filter(Objects::nonNull).findFirst().ifPresent(docRef -> {
@@ -257,7 +257,7 @@ public class ExtractionDecorator {
 
     private void extractData(final TaskContext parentContext,
                              final QueryKey queryKey,
-                             final AtomicLong extractionCount,
+                             final LongAdder extractionCount,
                              final ErrorConsumer errorConsumer) {
         taskContextFactory.childContext(parentContext, "Extraction Task", taskContext ->
                 securityContext.useAsRead(() -> {
@@ -293,7 +293,7 @@ public class ExtractionDecorator {
     private void extractEvents(final TaskContext taskContext,
                                final long streamId,
                                final Set<Event> events,
-                               final AtomicLong extractionCount,
+                               final LongAdder extractionCount,
                                final ErrorConsumer errorConsumer) {
         SearchProgressLog.increment(queryKey, SearchPhase.EXTRACTION_DECORATOR_FACTORY_CREATE_TASKS);
         SearchProgressLog.add(queryKey, SearchPhase.EXTRACTION_DECORATOR_FACTORY_CREATE_TASKS_EVENTS, events.size());
@@ -336,7 +336,7 @@ public class ExtractionDecorator {
                                     pipelineData);
                         });
 
-                        extractionCount.addAndGet(events.size());
+                        extractionCount.add(events.size());
 
                     } else {
                         // See if we can load the stream. We might get a StreamPermissionException if we aren't
@@ -358,7 +358,7 @@ public class ExtractionDecorator {
                         // Pass raw values to coprocessors that are not requesting values to be extracted.
                         for (final Event event : events) {
                             receiver.add(event.getValues());
-                            extractionCount.incrementAndGet();
+                            extractionCount.increment();
                         }
                     }
                 }

--- a/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/ClusterSearchTaskHandler.java
+++ b/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/ClusterSearchTaskHandler.java
@@ -37,7 +37,7 @@ import stroom.util.logging.LambdaLoggerFactory;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
 import javax.inject.Inject;
 
 class ClusterSearchTaskHandler {
@@ -49,8 +49,8 @@ class ClusterSearchTaskHandler {
     private final SecurityContext securityContext;
     private final ExecutorProvider executorProvider;
 
-    private final AtomicLong hitCount = new AtomicLong();
-    private final AtomicLong extractionCount = new AtomicLong();
+    private final LongAdder hitCount = new LongAdder();
+    private final LongAdder extractionCount = new LongAdder();
 
     private TaskContext taskContext;
 
@@ -143,10 +143,10 @@ class ClusterSearchTaskHandler {
         taskContext.info(() -> "" +
                 "Searching... " +
                 "found "
-                + hitCount.get() +
+                + hitCount.sum() +
                 " documents" +
                 " performed " +
-                extractionCount.get() +
+                extractionCount.sum() +
                 " extractions");
     }
 }

--- a/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/shard/IndexShardHitCollector.java
+++ b/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/shard/IndexShardHitCollector.java
@@ -16,19 +16,22 @@
 
 package stroom.search.impl.shard;
 
+import stroom.index.shared.IndexShard;
 import stroom.query.api.v2.QueryKey;
 import stroom.query.common.v2.SearchProgressLog;
 import stroom.query.common.v2.SearchProgressLog.SearchPhase;
 import stroom.task.api.TaskContext;
 import stroom.task.api.TaskTerminatedException;
+import stroom.util.NullSafe;
 import stroom.util.logging.LambdaLogger;
 import stroom.util.logging.LambdaLoggerFactory;
 
 import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Query;
 import org.apache.lucene.search.SimpleCollector;
 
 import java.io.IOException;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
 import java.util.function.Supplier;
 
 class IndexShardHitCollector extends SimpleCollector {
@@ -36,19 +39,26 @@ class IndexShardHitCollector extends SimpleCollector {
     private static final LambdaLogger LOGGER = LambdaLoggerFactory.getLogger(IndexShardHitCollector.class);
 
     private final TaskContext taskContext;
+    private final IndexShard indexShard;
     private final QueryKey queryKey;
+    private final Query query;
+
     //an empty optional is used as a marker to indicate no more items will be added
     private final DocIdQueue docIdQueue;
-    private final AtomicLong totalHitCount;
-    private final AtomicLong localHitCount = new AtomicLong();
+    private final LongAdder totalHitCount;
+    private final LongAdder localHitCount = new LongAdder();
     private int docBase;
 
     IndexShardHitCollector(final TaskContext taskContext,
                            final QueryKey queryKey,
+                           final IndexShard indexShard,
+                           final Query query,
                            final DocIdQueue docIdQueue,
-                           final AtomicLong totalHitCount) {
+                           final LongAdder totalHitCount) {
         this.taskContext = taskContext;
+        this.indexShard = indexShard;
         this.queryKey = queryKey;
+        this.query = query;
         this.docIdQueue = docIdQueue;
         this.totalHitCount = totalHitCount;
 
@@ -66,6 +76,7 @@ class IndexShardHitCollector extends SimpleCollector {
         // The interrupt status seems to be cleared somewhere is Lucene code so check here if we should terminate.
         if (taskContext.isTerminated()) {
             info(() -> "Quitting...");
+            LOGGER.debug("Quitting (terminated). {}, query term [{}]", this, query);
             throw new TaskTerminatedException();
         }
 
@@ -76,19 +87,20 @@ class IndexShardHitCollector extends SimpleCollector {
             SearchProgressLog.increment(queryKey, SearchPhase.INDEX_SHARD_SEARCH_TASK_HANDLER_DOC_ID_STORE_PUT);
             docIdQueue.put(docId);
             info(() -> "Found " + localHitCount + " hits");
+            LOGGER.trace("Collect called. {}, query term [{}]", this, query);
         } catch (final InterruptedException e) {
             info(() -> "Quitting...");
-            LOGGER.trace(e::getMessage, e);
+            LOGGER.debug("Quitting (interrupted). {}, query term [{}]", this, query);
             // Keep interrupting this thread.
             Thread.currentThread().interrupt();
             throw new TaskTerminatedException();
         } catch (final RuntimeException e) {
-            LOGGER.error(e::getMessage, e);
+            LOGGER.error("Error logging search progress: {}. {}", e.getMessage(), this, e);
         }
 
         // Add to the hit count.
-        localHitCount.getAndIncrement();
-        totalHitCount.getAndIncrement();
+        localHitCount.increment();
+        totalHitCount.increment();
     }
 
     private void info(final Supplier<String> message) {
@@ -99,6 +111,14 @@ class IndexShardHitCollector extends SimpleCollector {
     @Override
     public boolean needsScores() {
         return false;
+    }
+
+    @Override
+    public String toString() {
+        return "Query key: " + queryKey
+                + ", shard: " + NullSafe.get(indexShard, IndexShard::getId)
+                + ", shard hits: " + localHitCount.sum()
+                + ", total hits: " + NullSafe.getOrElse(totalHitCount, LongAdder::sum, -1);
     }
 }
 

--- a/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/shard/IndexShardHitCollector.java
+++ b/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/shard/IndexShardHitCollector.java
@@ -105,7 +105,7 @@ class IndexShardHitCollector extends SimpleCollector {
 
     private void info(final Supplier<String> message) {
         taskContext.info(message);
-        LOGGER.debug(message);
+        LOGGER.trace(message);
     }
 
     @Override

--- a/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/shard/IndexShardSearchFactory.java
+++ b/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/shard/IndexShardSearchFactory.java
@@ -35,7 +35,7 @@ import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
 import java.util.concurrent.atomic.AtomicInteger;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
 import javax.inject.Inject;
 import javax.inject.Provider;
 
@@ -75,7 +75,7 @@ public class IndexShardSearchFactory {
                                           final ExpressionOperator expression,
                                           final FieldIndex fieldIndex,
                                           final TaskContext parentContext,
-                                          final AtomicLong hitCount,
+                                          final LongAdder hitCount,
                                           final StoredDataQueue storedDataQueue,
                                           final ErrorConsumer errorConsumer) {
         SearchProgressLog.increment(task.getKey(), SearchPhase.INDEX_SHARD_SEARCH_FACTORY_SEARCH);

--- a/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/shard/IndexShardSearchTaskHandler.java
+++ b/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/shard/IndexShardSearchTaskHandler.java
@@ -49,7 +49,7 @@ import org.apache.lucene.util.Version;
 import java.io.IOException;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
 import javax.inject.Inject;
 
 public class IndexShardSearchTaskHandler {
@@ -83,7 +83,7 @@ public class IndexShardSearchTaskHandler {
                             final QueryKey queryKey,
                             final IndexShardQueryFactory queryFactory,
                             final String[] storedFieldNames,
-                            final AtomicLong hitCount,
+                            final LongAdder hitCount,
                             final int shardNumber,
                             final int shardTotal,
                             final long shardId,
@@ -96,8 +96,7 @@ public class IndexShardSearchTaskHandler {
                 taskContext.reset();
                 taskContext.info(() ->
                         "Searching shard " + shardNumber + " of " + shardTotal +
-                                " (id=" + shardId + ")");
-
+                                " (id=" + shardId + ")", LOGGER);
 
                 final IndexWriter indexWriter = getWriter(shardId);
 
@@ -145,7 +144,7 @@ public class IndexShardSearchTaskHandler {
     private void searchShard(final TaskContext parentContext,
                              final IndexShardQueryFactory queryFactory,
                              final String[] storedFieldNames,
-                             final AtomicLong hitCount,
+                             final LongAdder hitCount,
                              final IndexShardSearcher indexShardSearcher,
                              final ValuesConsumer valuesConsumer,
                              final ErrorConsumer errorConsumer) {
@@ -178,10 +177,17 @@ public class IndexShardSearchTaskHandler {
                                             final IndexShardHitCollector collector = new IndexShardHitCollector(
                                                     taskContext,
                                                     queryKey,
+                                                    indexShard,
+                                                    query,
                                                     docIdQueue,
                                                     hitCount);
 
                                             searcher.search(query, collector);
+
+                                            LOGGER.debug("Shard search complete. {}, query term [{}]",
+                                                    collector,
+                                                    query);
+
                                         } catch (final IOException e) {
                                             error(errorConsumer, e);
                                         }

--- a/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/shard/IndexShardSearchTaskHandler.java
+++ b/stroom-search/stroom-search-impl/src/main/java/stroom/search/impl/shard/IndexShardSearchTaskHandler.java
@@ -123,7 +123,7 @@ public class IndexShardSearchTaskHandler {
 
         } finally {
             if (indexShardSearcher != null) {
-                taskContext.info(() -> "Closing searcher for index shard " + shardId);
+                taskContext.info(() -> "Closing searcher for index shard " + shardId, LOGGER);
                 indexShardSearcher.destroy();
             }
         }

--- a/stroom-search/stroom-search-solr/src/main/java/stroom/search/solr/search/SolrClusterSearchTaskHandler.java
+++ b/stroom-search/stroom-search-solr/src/main/java/stroom/search/solr/search/SolrClusterSearchTaskHandler.java
@@ -37,7 +37,7 @@ import stroom.util.logging.LambdaLoggerFactory;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
 import javax.inject.Inject;
 
 class SolrClusterSearchTaskHandler {
@@ -49,8 +49,8 @@ class SolrClusterSearchTaskHandler {
     private final SecurityContext securityContext;
     private final ExecutorProvider executorProvider;
 
-    private final AtomicLong hitCount = new AtomicLong();
-    private final AtomicLong extractionCount = new AtomicLong();
+    private final LongAdder hitCount = new LongAdder();
+    private final LongAdder extractionCount = new LongAdder();
 
     private TaskContext taskContext;
     private QueryKey queryKey;
@@ -152,10 +152,10 @@ class SolrClusterSearchTaskHandler {
         taskContext.info(() -> "" +
                 "Searching... " +
                 "found "
-                + hitCount.get() +
+                + hitCount.sum() +
                 " documents" +
                 " performed " +
-                extractionCount.get() +
+                extractionCount.sum() +
                 " extractions");
     }
 }

--- a/stroom-search/stroom-search-solr/src/main/java/stroom/search/solr/search/SolrSearchFactory.java
+++ b/stroom-search/stroom-search-solr/src/main/java/stroom/search/solr/search/SolrSearchFactory.java
@@ -22,7 +22,7 @@ import org.apache.solr.client.solrj.SolrQuery;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
 import javax.inject.Inject;
 
 public class SolrSearchFactory {
@@ -60,7 +60,7 @@ public class SolrSearchFactory {
                                           final StoredDataQueue storedDataQueue,
                                           final ErrorConsumer errorConsumer,
                                           final TaskContext parentContext,
-                                          final AtomicLong hitCount,
+                                          final LongAdder hitCount,
                                           final String dateTimeLocale) {
         // Make sure we have been given a query.
         if (query.getExpression() == null) {

--- a/stroom-search/stroom-search-solr/src/main/java/stroom/search/solr/search/Tracker.java
+++ b/stroom-search/stroom-search-solr/src/main/java/stroom/search/solr/search/Tracker.java
@@ -5,25 +5,25 @@ import stroom.util.logging.LambdaLoggerFactory;
 
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
 
 class Tracker {
 
     private static final LambdaLogger LOGGER = LambdaLoggerFactory.getLogger(Tracker.class);
 
-    private final AtomicLong hitCount;
+    private final LongAdder hitCount;
     private final CountDownLatch completed = new CountDownLatch(1);
 
-    Tracker(final AtomicLong hitCount) {
+    Tracker(final LongAdder hitCount) {
         this.hitCount = hitCount;
     }
 
     long getHitCount() {
-        return hitCount.get();
+        return hitCount.sum();
     }
 
     void incrementHitCount() {
-        hitCount.incrementAndGet();
+        hitCount.increment();
     }
 
     boolean awaitCompletion(final long timeout, final TimeUnit unit) throws InterruptedException {

--- a/stroom-task/stroom-task-api/src/main/java/stroom/task/api/TaskContext.java
+++ b/stroom-task/stroom-task-api/src/main/java/stroom/task/api/TaskContext.java
@@ -17,12 +17,30 @@
 package stroom.task.api;
 
 import stroom.task.shared.TaskId;
+import stroom.util.NullSafe;
+
+import org.slf4j.Logger;
 
 import java.util.function.Supplier;
 
 public interface TaskContext extends Terminator {
 
     void info(Supplier<String> messageSupplier);
+
+    /**
+     * Set the messageSupplier on the taskContext and if logger has debug enabled
+     * then log the message as well.
+     */
+    default void info(final Supplier<String> messageSupplier, final Logger logger) {
+
+        info(messageSupplier);
+
+        if (logger != null && logger.isDebugEnabled() && messageSupplier != null) {
+            logger.debug("TaskId: {}, info: {}",
+                    NullSafe.get(getTaskId(), TaskId::getId),
+                    messageSupplier.get());
+        }
+    }
 
     /**
      * Get the task id of this context.

--- a/stroom-task/stroom-task-impl/src/main/java/stroom/task/impl/TaskContextImpl.java
+++ b/stroom-task/stroom-task-impl/src/main/java/stroom/task/impl/TaskContextImpl.java
@@ -21,6 +21,9 @@ import stroom.task.api.TaskContext;
 import stroom.task.api.TaskTerminatedException;
 import stroom.task.api.TerminateHandler;
 import stroom.task.shared.TaskId;
+import stroom.util.NullSafe;
+
+import org.slf4j.Logger;
 
 import java.util.Collections;
 import java.util.Objects;
@@ -62,6 +65,19 @@ public class TaskContextImpl implements TaskContext {
     @Override
     public void info(final Supplier<String> messageSupplier) {
         this.messageSupplier = messageSupplier;
+    }
+
+    @Override
+    public void info(final Supplier<String> messageSupplier, final Logger logger) {
+        this.messageSupplier = messageSupplier;
+
+        if (logger != null && logger.isDebugEnabled() && messageSupplier != null) {
+            logger.debug("TaskId: {}, user: {}, task name: {}, info: {}",
+                    NullSafe.get(taskId, TaskId::getId),
+                    NullSafe.get(userIdentity, UserIdentity::getId),
+                    name,
+                    messageSupplier.get());
+        }
     }
 
     @Override

--- a/stroom-task/stroom-task-impl/src/main/java/stroom/task/impl/TaskContextImpl.java
+++ b/stroom-task/stroom-task-impl/src/main/java/stroom/task/impl/TaskContextImpl.java
@@ -25,7 +25,9 @@ import stroom.util.NullSafe;
 
 import org.slf4j.Logger;
 
+import java.util.ArrayList;
 import java.util.Collections;
+import java.util.List;
 import java.util.Objects;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
@@ -72,11 +74,30 @@ public class TaskContextImpl implements TaskContext {
         this.messageSupplier = messageSupplier;
 
         if (logger != null && logger.isDebugEnabled() && messageSupplier != null) {
-            logger.debug("TaskId: {}, user: {}, task name: {}, info: {}",
-                    NullSafe.get(taskId, TaskId::getId),
+            logger.debug("Task stack: {}, user: {}, task: {}, info: {}",
+                    getTaskHierarchy(),
                     NullSafe.get(userIdentity, UserIdentity::getId),
                     name,
                     messageSupplier.get());
+        }
+    }
+
+    private String getTaskHierarchy() {
+        if (taskId == null) {
+            return "";
+        } else {
+            final List<String> taskIds = new ArrayList<>();
+            TaskId currTaskId = this.taskId;
+            while (currTaskId != null) {
+                taskIds.add(currTaskId.getId());
+                currTaskId = currTaskId.getParentId();
+                // In case we get a huge hierarchy
+                if (currTaskId != null && taskIds.size() >= 10) {
+                    taskIds.add("...");
+                    break;
+                }
+            }
+            return String.join(" < ", taskIds);
         }
     }
 

--- a/unreleased_changes/20220519_172144_292__2874.md
+++ b/unreleased_changes/20220519_172144_292__2874.md
@@ -1,0 +1,24 @@
+* Issue **#2874** : Add debug logging for shard writing/reading.
+
+
+```sh
+# ********************************************************************************
+# Issue title: Exception thrown for Index shard.
+# Issue link:  https://github.com/gchq/stroom/issues/2874
+# ********************************************************************************
+
+# ONLY the top line will be included as a change entry in the CHANGELOG.
+# The entry should be in GitHub flavour markdown and should be written on a SINGLE
+# line with no hard breaks. You can have multiple change files for a single GitHub issue.
+# The  entry should be written in the imperative mood, i.e. 'Fix nasty bug' rather than
+# 'Fixed nasty bug'.
+#
+# Examples of acceptable entries are:
+#
+#
+# * Issue **123** : Fix bug with an associated GitHub issue in this repository
+#
+# * Issue **namespace/other-repo#456** : Fix bug with an associated GitHub issue in another repository
+#
+# * Fix bug with no associated GitHub issue.
+```

